src/Umbraco.Infrastructure/Services/Implement/ContentService.cs

```diff
@@ -773,7 +773,7 @@ public OperationResult Save(IContent content, int userId = Cms.Core.Constants.Se
 
                 if (raiseEvents)
                 {
-                    scope.Events.Dispatch(Saved, this, saveEventArgs.ToContentSavedEventArgs(), nameof(Saved));
+                    scope.Notifications.Publish(new SavedNotification<IContent>(content, evtMsgs));
```

Comment:
@kjac There's one potential issue here with these changes. The reason why in many cases there's an extensions like `ToContentSavedEventArgs()` is to ensure that the `AdditionalData` flows and is the same instance. This concept is used in a few places to be able to share data between the "Saving" (before) and "Saved" (after) events. 

This is just an example but this concept is used in most event raising places. 

Perhaps you've already addressed that somewhere in this code? I haven't had a deep dive into this, just a quick scan.

Reply:
@Shazwazza yeah I know. I forgot to add that to the description - will update it. I have chosen to remove both `EventState` and `AdditionalData` in this implementation. I really dislike how the "ed" and "ing" events are tightly coupled like that... it feels fragile. In my opinion, event state should be a matter for the notification subscribers to handle in their specific implementations. 

If you look at the `RedirectTrackingHandler` rewrite in this PR, I have solved it there using `IRequestCache` as a temporary storage mechanism between the "ed" and "ing" events.

With this PR, `EventState` is no longer used anywhere in the Umbraco core, and even before this PR, `AdditionalData` is only ever used once (by `ModelsBuilderNotificationHandler`). 

Thoughts?

Reply:
Unfortunately that isn't gonna work. Request cache is only available during an http request. That will be null when running in a non request which happens all the time with services. We need to be able to pass state between events and cannot rely on an external state machine. That is even more fragile and is exactly why this has been done, else people will hack some static state bag. Happy to make this nicer some how but the event instance passed for before and after events will need to be able to carry state

Reply:
Makes sense. This is not the hill I will die on, and I'll be happy to reintroduce the event state if that's the best solution. It just feels a bit strange to me to clone state between two instances of different classes in order to pass it between events.

Another idea could be to have a root scoped cache implementation on `Scope`. This could also be used by others throughout the current scope, thus not bound to events (notifications).

I dunno... just want to be sure we're doing state handling the best possible way here, since we'll have a hard time changing it later on. Again, happy to redo the event state on the cancelable notifications if that's the best approach.

Reply:
Totally happy to re-think how we do it! The 'AdditionalData' thing was just added as a means to an end without breaking changes. The Scope idea is interesting however I'm unsure that will work with the 'post' events since those execute when the scope is disposed and since we are not passing Scope around in these events we'd have to rely on the Ambient Scope which if acquired in the 'post' events may not be the same Scope instance (or in the hierarchy chain) as the one used in the service.

Reply:
@Shazwazza I just had a go at it. You're absolutely right, it won't fly. I'll revert to reintroducing state on the notifications. 👍

Reply:
Alright. I have introduced a new interface `IStatefulNotification`, which defines a `State` dictionary, and extension methods in `NotificationExtensions` to copy state between stateful notifications. 

`RedirectTrackingHandler` now utilizes this state to track redirects.

Note that both content and media service as-is have a few events that do not carry state between "ing" and "ed". I have tried to retain this behavior for the time being.